Feature(journal core): Introduce core posting journal

sfount opened the pull request:
This pull request introduces a proposal for how to structure the core client side posting journal utility. 

Many core features of Angular UI Grid have been implemented with a focus on how adding additional features will not greatly increase complexity. 

`JournalController` - Configures the base UI grid and initialises the configuration services 
`Journal-X-Service` - Journal features (sorting, pagination, filtering etc.) are separated into external services that can be found in _js/services/journal/'_, this allows the majority of the complex configuration and utility methods to be defined independent of each other
`TransactionService` - This is a very basic introduction to how the posting journal can interact with the server API end point; right now it provides methods for creating test data and also fetching records from the server. This will be updated in future development to factor in searching, limiting transactions, etc. 

`journal.spec.js` - Base testing framework, this should be extended to match every feature added
`journal.page.js` - Journal page object, this is used to interact with the journal page during tests
### Posting Journal default state (transaction view)

![posting journal default state](http://i.imgur.com/VBlgkGv.jpg)
### Selecting a transaction

![selecting a transaction](http://i.imgur.com/i8ndVyw.jpg)
### Basic feature demonstration
- Column moving 
- Column resizing 
- Row selection 
- Column hiding
- Pagination
- Grouping (by transaction)

![basic feature demonstration](http://i.imgur.com/6H48JWm.jpg)

sfount committed:
Prototype initial core posting journal files

This commit introduces the initial files for the 2.x client side
posting journal. It proposes a structure for how the core UI grid
should be configured and sets up the testing framework.

Testing - A page object is used to abstract methods on the page
components (primarily the grid), currently this provides a method to
verify the total number of rows that can be accesed by the test runner
(journal.spec.js). This method is taken from the test utilities provided
by Angular UI Grid, this is a fully featured set of tools however the
actual expect behaviour is configured for a different testing framework.

Service Infrastructure - A service 'TransactionService' has been
created to interface with the server journal end point and provide any
utility methods required to manipulate transactions. Seperating models
and functionality into services will be a key aspect of how the
controller and module achieves modularity.

Routing - app.js is updated such that /#/journal points to the new 2.x
journal files. Once the updated posting journal is stable it will
completely replace the previous implementation; this should be reflected
in the file structure.

sfount committed:
(Chore) Mock transaction data algorithm

This commit adds methods to mock development data required for testing
and validating grid functionality. These methods will be depricated once
the system is able to produce the data through its core components.

Journal column definitions are updated to match something like the current
posting journal database; there are a number of changes that will have to be made;
inv_po_id -> reference
currency_symbol -> currency
cc -> cost_center
pc -> cost_center

Updated tests to reflect new mock data.

The mock transaction algorithm requires refining to handle the following
cases:
* Balanced transactions given a total cost
* Sensible account assignment with each cost
* Assigning debitors and creditors
* Assigning reference documents

Once this is completed a static set of test data can be generated and
used to populate models/test/data.sql

sfount committed:
Propose modular service based architecture - Sort

This commit proposes a service based approach for each of the utilities the
journal will require. In this proposal sorting is implemented by creating the
JournalSortingService. This service is responsible for
- Configuring the global gridOptions object
- Providing helper functions for complex sorting

TransactionService was renamed to journal/TransactionService, this service
will still be available to any controllers that require however the current
implementation is directly intended for the posting journal.

sfount committed:
Implemented additional feature service (Grouping)

This commit adds basic grouping into the core posting journal. Grouping
currently sets default values and extends the grid with two basic
features:
- Selecting a group header selects each of the children for that row
- Grid groups are expanded on initialise

The sort algorithm was updated to support grouping on transaction IDs.
Angular UI Grid does not pass an row object to the sorting method for
group header rows (because it is just a UI object - there is no
underlying data for these rows). Previously the sort worked by comparing
the rows `reference` attribute - the algorithm now parses this from the
transaction ID string if it is a group header (the only information
passed to headers).

Proposed onRegisterApi callback method allows services to register a
callback and retrieve the API; once they have completed that they call
any method that was registrered before them.

Added:
- Column resizing
- Column moving
- Column selection

sfount committed:
Implement pagination service

This commit adds the files required for the journal pagination service,
providing a number of global configurations as well as helper methods to
settup correct pagination.

JournalPagination can be configured to specify if the service should use
external (custom) pagination (as defined in the
JournalPaginationService) or to use the default Angular UI Grid
pagination. Custom pagination (driven by the method `fetchPages`)
respects transactions and will not split a transaction across multiple
pages; this means the page size is dependent on the current page of
trnnsactions.

Currently isExternalPagination is set to false; until the jounral
requires features from the custom pagination this will be set - there
are many corner cases to consider with pagination (sorting, selection,
cacheing) that will require a very careful tested design.

- Updated column definitions go group (and sort) on transactions by
  default. This gives the core view into the posting journal.

DedrickEnc commented on client/src/js/app.js:
```diff
@@ -178,18 +179,19 @@ function bhimaconfig($routeProvider) {
   /* journal routes */
 
-  .when('/journal', {
+  .when('/1.X-journal', {
```
Do we need to keep these routes?

sfount replied:
We do not - I left these in case anyone wanted to reference the previous posting journal structure; this can easily be done using git though!

jniles commented on client/src/js/services/journal/JournalGrouping.js:
```diff
@@ -0,0 +1,86 @@
+ 
+  /**
+   * This method enables group level selections when the posting jounral is grouped
+   * according to a column. It first checks to ensure the row that changed is a row 
```
Minor typo: jounral  :arrow_right: journal.

Slightly more serious typo: "It first checks to ensure the row that changed is a row" should probably read "It first checks to ensure that the row that changed is a **header** row and then selects the row's children."

sfount replied:
:sob: That should be updated before being merged - good catch.

sfount committed:
(Chore) Fixed typos and removed legacy routes

This commit removes the old 1.x journal routes. These can be recovered
by checking the files out of git history. Minor typo fixes in
JournalGrouping (row -> header).
